runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -514,20 +514,26 @@ function LanguageTree:register_cbs(cbs)
 end
 
 ---@private
-local function tree_contains(tree, range)
-  local start_row, start_col, end_row, end_col = tree:root():range()
-  local start_fits = start_row < range[1] or (start_row == range[1] and start_col <= range[2])
-  local end_fits = end_row > range[3] or (end_row == range[3] and end_col >= range[4])
+local function region_contains(region, range)
+  for _, node in ipairs(region) do
+    local start_row, start_col, end_row, end_col = node:range()
```

Comment:
I think `region_contains` with the loop not in this function would it more general an easier to understand what it does (this is essentially `ts_util.is_node_in_range` that also works for range in ranges). I think the same could stay the `region_contains` even when the loop gets pulled out into `Language:tree`. The regions are actually only ranges and not the table containing the regions for one tree.

 I think regions are not necessarily nodes but can also be ranges, in case `offset!` cut a bit off the region.

Tests would nice, since it's always difficult to remember what regions actually are. I for example don't remember why they are doubly nested but I think this has to do with how @steelsojka implemented "combined" that you can actually have multiple trees of one injection in the same document (what upstream tree-sitter does not do). Rethinking that then this makes sense.

For incremental parsing, we should probably get rid of the outer level of regions and just adapt the `set_included_regions` without ever creating additional trees.

Reply:
So, at the moments `_regions = {{region}}` where region = `node | range`. It would probably be simpler when we would just have `{region}` with only one tree instead of `_trees`

Reply:
> I think `region_contains` with the loop not in this function would it more general an easier to understand what it does (this is essentially `ts_util.is_node_in_range` that also works for range in ranges). I think the same could stay the `region_contains` even when the loop gets pulled out into `Language:tree`. The regions are actually only ranges and not the table containing the regions for one tree.

Oh yeah, that's nicer :+1:

> Tests would nice, since it's always difficult to remember what regions actually are.

What kinds of tests would you like to see there? I'll definitely try to add one for the case that is fixed with this PR, but I'm not sure what to do about the regions.

> I for example don't remember why they are doubly nested but I think this has to do with how @steelsojka implemented "combined" that you can actually have multiple trees of one injection in the same document

From `:h LanguageTree:set_included_regions` it sounds like it's the other way around, ie. one tree for multiple injected regions?

Reply:
> From :h LanguageTree:set_included_regions it sounds like it's the other way around, ie. one tree for multiple injected regions?

Yes, this is how upstream tree-sitter does it. Did you check whether a region can also be a range instead of a node?

Reply:
Yes, it's possible (and actually done) to construct a region from a range.